tests/fuzzer/flatbuffers_scalar_fuzzer.cc

```diff
@@ -0,0 +1,380 @@
+#include <stddef.h>
+#include <stdint.h>
+#include <algorithm>
+#include <clocale>
+#include <memory>
+#include <regex>
+#include <string>
+
+#include "flatbuffers/idl.h"
+
+#define flags_scalar_type 0x0F  // type of scalar value
+#define flags_quotes_kind 0x10  // quote " or '
+#define flags_json_bracer 0x20  // {} or []
+#define flags_0x40 0x40         // reserved flag
+#define flags_clocale 0x80      // change default C-locale
+
+#define TEST_OUTPUT_LINE(...)     \
+  {                               \
+    fprintf(stderr, __VA_ARGS__); \
+    fprintf(stderr, "\n");        \
+  }
+
+static void TestFail(const char *expval, const char *val, const char *exp,
+                     const char *file, int line) {
+  TEST_OUTPUT_LINE("VALUE: \"%s\"", expval);
+  TEST_OUTPUT_LINE("EXPECTED: \"%s\"", val);
+  TEST_OUTPUT_LINE("TEST FAILED: %s:%d, %s", file, line, exp);
+  assert(0);
+}
+
+static void TestEqStr(const char *expval, const char *val, const char *exp,
+                      const char *file, int line) {
+  if (strcmp(expval, val) != 0) { TestFail(expval, val, exp, file, line); }
+}
+
+template<typename T, typename U>
+void TestEq(T expval, U val, const char *exp, const char *file, int line) {
+  if (U(expval) != val) {
+    TestFail(flatbuffers::NumToString(expval).c_str(),
+             flatbuffers::NumToString(val).c_str(), exp, file, line);
+  }
+}
+
+#define TEST_EQ(exp, val) TestEq(exp, val, #exp, __FILE__, __LINE__)
+#define TEST_NOTNULL(exp) TestEq(exp == NULL, false, #exp, __FILE__, __LINE__)
+#define TEST_EQ_STR(exp, val) TestEqStr(exp, val, #exp, __FILE__, __LINE__)
+
+static void BreakSequence(std::string &s, const char *subj, char repl) {
+  size_t pos = 0;
+  while (pos = s.find(subj, pos), pos != std::string::npos) s.at(pos) = repl;
+}
+
+static std::string strip_string(const std::string &s, const char *pattern,
+                                size_t *pos = nullptr) {
+  if (pos) *pos = 0;
+  // leading
+  auto first = s.find_first_not_of(pattern);
+  if (std::string::npos == first) return "";
+  if (pos) *pos = first;
+  // trailing
+  auto last = s.find_last_not_of(pattern);
+  assert(last < s.length());
+  assert(first <= last);
+  return s.substr(first, last - first + 1);
+}
+
+class RegexMatcher {
```

Comment:
Maybe some comments on what all the new code in this file does and why it is needed?